tests/unit/autocomplete/test_indexer.py

```diff
@@ -0,0 +1,188 @@
+                                    'filters', 'list', nargs='+'),
+                            }
+                        ),
+                        'run-instances': DummyCommand('run-instances'),
+                    }
+                ),
+                's3': DummyCommand(
+                    command_name='s3',
+                    subcommand_table={
+                        'list-objects': DummyCommand('list-objects'),
+                        'put-object': DummyCommand('put-object'),
+                    }
+                )
+            }
+        )
+        # Ideally we just use ':memory:', but that's for
+        # a specific sqlite3 connection so we'd have to
+        # update our interfaces to return/accept a db connection.
+        # I'd prefer not to have db connections as part of the
+        # indexing interface.
+        self.tempfile = tempfile.NamedTemporaryFile('w')
+        self.db_conn = indexer.DatabaseConnection(self.tempfile.name)
+        self.indexer = indexer.ModelIndexer(self.db_conn)
+        self.query = model.ModelIndex(self.tempfile.name)
+
+    def tearDown(self):
+        self.tempfile.close()
+
+    def test_can_retrieve_top_level_commands(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.command_names(lineage=['aws'])),
+            set(['ec2', 's3'])
+        )
+
+    def test_can_retrieve_operation_names(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.command_names(lineage=['aws', 'ec2'])),
+            set(['describe-instances', 'run-instances']),
+        )
+
+    def test_can_retrieve_global_params(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.arg_names(lineage=[], command_name='aws')),
+            set(['region', 'endpoint-url']),
+        )
+
+    def test_can_retrieve_service_params(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.arg_names(lineage=['aws', 'ec2'],
+                                     command_name='describe-instances')),
+            set(['instance-ids', 'filters']),
+        )
+
+    def test_can_retrieve_correct_commands_when_shadowed(self):
+        # Suppose that 's3' had a 'describe-instances' operation.  We
+        # should be able to differentiate between this command and the
+        # one in ec2 via the ``parent`` arg.
+        s3_commands = self.aws_command.subcommand_table['s3'].subcommand_table
+        s3_commands['describe-instances'] = DummyCommand('describe-instances')
+        self.indexer.generate_index(self.aws_command)
+        # The 'ec2' version has several params in its arg table, but this
+        # 's3' version should have no params because we didn't add an arg
+        # table.
+        self.assertEqual(
+            set(self.query.arg_names(lineage=['aws.s3'],
```

Comment:
I think we need to update this test case as well:
```suggestion
            set(self.query.arg_names(lineage=['aws', 's3'],
```